### PR TITLE
Fix #3263: Fix failing in `utsnameOps`

### LINKED ISSUE #3263: posixlib: ClassCastException occurred in `utsnameOps`
Hi team, I recently started using Scala Native and have encountered the following unexpected behavior:

## Env
Scala Native: 0.4.12
Scala: 2.13.10

## Reproducer

```scala
import scala.scalanative.unsafe._
import scala.scalanative.posix.sys.utsname._
import scala.scalanative.posix.sys.uname._

object Test {
  def main(args: Array[String]): Unit = {
    val u: Ptr[utsname] = stackalloc[utsname]()

    if (uname(u) == -1)
      throw new RuntimeException("uname err")

    // this causes error
    println(s"${u.sysname}")
  }
}
```
## Output

```
java.lang.ClassCastException: scala.scalanative.unsafe.CArray cannot be cast to scala.scalanative.unsafe.Ptr
        at java.lang.StackTrace$.$anonfun$currentStackTrace$1(Unknown Source)
        at java.lang.StackTrace$$$Lambda$2.apply(Unknown Source)
        at scala.scalanative.unsafe.Zone$.apply(Unknown Source)
        at java.lang.StackTrace$.currentStackTrace(Unknown Source)
        at java.lang.Throwable.fillInStackTrace(Unknown Source)
        at scala.scalanative.runtime.package$.throwClassCast(Unknown Source)
        at scala.scalanative.posix.sys.uname$utsnameOps.sysname(Unknown Source)
        at Test$.main(Unknown Source)
        at Test.main(Unknown Source)
        at <none>.main(Unknown Source)
[error] java.lang.RuntimeException: Nonzero exit code: 1
[error]         at scala.sys.package$.error(package.scala:30)
[error]         at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$23(ScalaNativePluginInternal.scala:292)
[error]         at scala.Option.foreach(Option.scala:407)
[error]         at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$21(ScalaNativePluginInternal.scala:292)
[error]         at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$21$adapted(ScalaNativePluginInternal.scala:269)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] stack trace is suppressed; run last Compile / run for the full output
[error] (Compile / run) Nonzero exit code: 1
```

### PR DESCRIPTION
**Pull request by mox692**
Closes [#3263](https://github.com/scala-native/scala-native/issues/3263).

### CODE REVIEW

File: posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala [hunk 1]
```diff
@@ -13,10 +13,10 @@ object utsname {
...
-    def sysname = fromCString(c._1.asInstanceOf[Ptr[CChar]])
-    def nodename = fromCString(c._2.asInstanceOf[Ptr[CChar]])
-    def release = fromCString(c._3.asInstanceOf[Ptr[CChar]])
-    def version = fromCString(c._4.asInstanceOf[Ptr[CChar]])
-    def machine = fromCString(c._5.asInstanceOf[Ptr[CChar]])
+    def sysname = fromCString(c.at1.asInstanceOf[Ptr[CChar]])
+    def nodename = fromCString(c.at2.asInstanceOf[Ptr[CChar]])
+    def release = fromCString(c.at3.asInstanceOf[Ptr[CChar]])
+    def version = fromCString(c.at4.asInstanceOf[Ptr[CChar]])
+    def machine = fromCString(c.at5.asInstanceOf[Ptr[CChar]])
```
[the remark this answers is not in the document]
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala:20]: I would say that the POSIX design here is a bit lacking.  I am not sure about other code that may be similar either off the top of my head. @LeeTibbert, Do you have any ideas?

[the remark this answers is not in the document]
Comment (mox692) [reply posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala:20]: So the point here is whether it is ok or not to implicitly provide Ops that convert to String(where user would normally expect an Array of Char), right?

[the remark this answers is not in the document]
Comment (armanbilge) [reply posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala:20]: Yeah, in my opinion the point of the Ops is to replace the `_1`, `_2`, ... stuff with the actual struct names for readability. Besides that, it should not do any other conversion—this is supposed to be a raw binding, not a user-friendly interface.

[the remark this answers is not in the document]
Comment (mox692) [reply posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala:20]: @LeeTibbert 
Thank you for the detailed comment.

I also think it's not good to implicitly convert types with ops, so I will change `utsnameOps` to return CArray like other parts.

Of course, I will also update the docs.
I found the 0.5.0 release docs [here](https://github.com/scala-native/scala-native/blob/main/docs/changelog/0.5.0.md) and [build method](https://github.com/scala-native/scala-native/blob/main/docs/contrib/build.rst#locally-publish-docs), so should I summarize the discussion here in this docs and include this PR?

To summarize: I'll make this PR for 0.5.0 instead of 0.4.x, change the behavior of the Ops, and update the docs to reflect this change. Is that okay?


(I apologize in advance for any delays in my response as it's a long holiday week in my country.)

[the remark this answers is not in the document]
Comment (LeeTibbert) [reply posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala:20]: re: long holiday weekend

Maybe I should move to your country, at least for the weekend....

Re: PR

Yes, I think the plan you describe is good.   `docs/changelog/0.5.0.md` looks like
the right file.  That is currently a rough draft document so that the wording &
placement need only be a first approximation: fit for purpose and no obvious
blunders/misspellings but not finely tuned wordsmithing.

If the `makedocs` works for you, great.  I have had problems with it; I
think with the installing Sphinx part.  I might assume the presence
of a (python) .mumble file or such which I do not have on my system.

IIRC I have an open Issue here, which I have never gotten the time to fix.

I know sphinx is needed for .rst (rich structured text) files. I do not know
if all of it is needed for .md (markdown) files.

Not to lead anyone astray... I usually change directory to the docs directory
and then run "make html" on the command line. I then use a "file://" URL
with the "//" replaced by  two slashes and the fully qualified (which add a third slash)
name of the _build/html directory (or a specific file under it).

Sorry if this quicky tutorial is describing stuff you already know, perhaps better than I.
Trying to save you time, and importantly frustration/pain.

Contributors are precious; they need encouragement.

[the remark this answers is not in the document]
Comment (mox692) [reply posixlib/src/main/scala/scala/scalanative/posix/sys/utsname.scala]: I’m back :)

@LeeTibbert
Thank you for the detailed explanation, and I was able to build the docs without any issues.

So I modified the function to return a CArray instead of a String, and added some description to 0.5.0 docs about this change.

By the way, I’m wondering if I should fix the behavior of Ops in version 0.4.x.
Should we make a change that only fixes the bug while keeping api to return a string?
(And would that probably be a separate PR?)

**Review by WojciechMazur — APPROVED** [review]

File: unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/UtsnameTest.scala [hunk 2]
```diff
@@ -21,39 +21,39 @@ class UtsnameTest {
       0
     )
 
-    val sysname = u.at1.asInstanceOf[Ptr[CChar]]
-    val nodename = u.at2.asInstanceOf[Ptr[CChar]]
-    val release = u.at3.asInstanceOf[Ptr[CChar]]
-    val version = u.at4.asInstanceOf[Ptr[CChar]]
-    val machine = u.at5.asInstanceOf[Ptr[CChar]]
+    val sysname = u._1.asInstanceOf[CArray[Byte, _256]]
```
Comment (LeeTibbert) [inline unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/UtsnameTest.scala:24]: Thank you for this iteration.

posixlib has one foot in C land and the other in Java Land,
so things get complex.

I suggest leaving these as `Ptr[CChar]]`, in general and especially in Test code.  That is, treat them as CStrings
and not as CArray (of known size).

The Open Group 2018 standard says:
```
The character arrays are of unspecified size, but the data stored in them shall be terminated by a null byte.
```

Now in Test code, there may be some cause to set a maximum size tested, say 1024 or such, to ensure
that the OS did not return an "unreasonable" or non-null-terminated string.  I will have to look at the rest of the code
in this file.

Am I missing a place where the magic _256 is defined.  The
OS could have returned larger than that. I think it is 
OS dependent.

Comment (LeeTibbert) [reply unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/UtsnameTest.scala:24]: I think I would also use the Ops name, rather than _1, just
to make sure they get exercised.  If something _can_ go
wrong, it will.

Then again, I am not the one writing this PR and must 
defer to the author.

Comment (mox692) [reply unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/UtsnameTest.scala:24]: Thank you for the feedback and for addressing it in another issue. I will check it out when I have time.

Comment (LeeTibbert) [reply unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/UtsnameTest.scala:24]: mox692.   Any improvements are welcome, just want to re-assure you that just because you touched it....

If you have time and motivation, I think the "short fix" (change to _257 and give Open Group, man 7, and
issue) is cost effective.  Leaving the Issue open pending the proper long term fix may be useful, so that
the "proper" fix is not forgotten.

I hope you had a good holiday weekend.
